### ISSUE TITLE
Question - Specific User & Item combination score?

### ISSUE THREAD

**Issue by dataedgehungary** [microsoft/Product-Recommendations#26]
Hi All!

I am trying to find a solution for the following: 

I have transactional data connected to users (loyalty card) from supermarkets. Roughly 30.000 items and 100.000 users with over a million transactional data.
Now, to train the model and get random recommendations is not a problem, the solution is great for that, but I cannot get my head around how to get individual scores for user & item combinations.

Basically, I have a weekly catalogue of around 50 items on sale. I would like to display these "coupons" in descending order in our mobile application where the user is signed in, based on his Recommendations Score for each item. 

Is there a way to input user id & product id and get the score?

Thanks in advance!
Mark

**Comment by dataedgehungary** [comment]
Alternatively, can I include a list of item ids, to recommend from? Obviously using the full catalogue is necessary to evaluate user-to-user similarity, but for the actual recommendation itself, can I limit the items somehow?

**Comment by natinimni** [comment]
Hi Mark,

The recommendation API has a few endpoints (“usage item” refers to the <item-id, timestamp, event> triplet, although only the item-id is mandatory):
1.	[Get Recommendations by multiple usage item](https://github.com/Microsoft/Product-Recommendations/blob/master/doc/api-reference.md#get-personalized-recommendations) - the service will return recommendation that match the set of input item ids
2.	[Get Recommendations by item id](https://github.com/Microsoft/Product-Recommendations/blob/master/doc/api-reference.md#get-item-to-item-recommendations) – a simplified API but essentially the same as option 1, only using a single item id as the input andg “now” as its timestamp
3.	Get Recommendations by user id – the same as option 1 (when providing a **user-id** as input), only the set of usage items is the “user history”, i.e. the specific user’s usage items used when the model was trained
4.	Get Recommendations by user id and usage items – the same as option 3, with the addition of the input usage items
*Note that if timestamps are provided, a more recent usage item would have greater impact on recommendation.

I hope this helps, it seems you can use options 3 or 4 for your scenario.

Thanks,
Nati

**Comment by dataedgehungary** [comment]
Good to hear, but can you write an example request for me, that returns the Score for let's say UserID: 0001 and ItemId: 10001?




Lekeny Márk


ügyvezető


DataEdge Kft.




+36 30 436 0799


mlekeny@dataedge.hu







On Tue, Jul 24, 2018 at 6:46 PM +0200, "Nati Nimni" <notifications@github.com> wrote:












Hi Mark,


The recommendation API has a few endpoints (“usage item” refers to the <item-id, timestamp, event> triplet, although only the item-id is mandatory):

Get Recommendations by multiple usage item - the service will return recommendation that match the set of input item ids
Get Recommendations by item id – a simplified API but essentially the same as option 1, only using a single item id as the input andg “now” as its timestamp
Get Recommendations by user id – the same as option 1 (when providing a user-id as input), only the set of usage items is the “user history”, i.e. the specific user’s usage items used when the model was trained
Get Recommendations by user id and usage items – the same as option 3, with the addition of the input usage items

*Note that if timestamps are provided, a more recent usage item would have greater impact on recommendation.



I hope this helps, it seems you can use options 3 or 4 for your scenario.


Thanks,

Nati

**Comment by LuisCabrer** [comment]
Mark, 

Nati and I discussed in a bit more detail -- and we recommend a bit of a diffrent strategy:

We don't have a single API that will return you the affinity between a user and a product. 

That said, there is a bit of a hack you can do. You could simply ask for a very large number of recommendations for the user of interest, and then check if any of the items that you want to promote are in the result. If they are, then you can bump those to the top of the promotion list.

By the way, examples of calls like this one are documented in the links Nati sent you earlier.

POST https://<service_name>.azurewebsites.net/api/models/e16198c0-3a72-4f4d-b8ab-e4c07c9bccdb/recommend?userId=0001 
x-api-key: your_api_key


Note that if you want items in the transaction history of the user to be returned in the recommendations, you need to set the  allowSeedItemsInRecommendations parameter to true when you first create the model.

Thanks!
-Luis

**Comment by dataedgehungary** [comment]
Hi Luis,




Thanks for the super quick response :) I was thinking of that solution as well, requesting as many recommendations as possible, seems like a way around.




I guess it has some implications on the speed of execution?




I will give it a try tomorrow. I'm thinking whether it is better if I iterate through all the users with let's say 30000 requested recommendations once a week or so, and save that data to a database, then I expose that to our own API and just run a select querry, when the application calls it.




Anyway, I will give it a try and keep you posted.




Best Regards,




Lekeny Márk


ügyvezető


DataEdge Kft.




+36 30 436 0799


mlekeny@dataedge.hu







On Tue, Jul 24, 2018 at 7:49 PM +0200, "Luis Cabrera" <notifications@github.com> wrote:












Mark,


Nati and I discussed in a bit more detail -- and we recommend a bit of a diffrent strategy:


We don't have a single API that will return you the affinity between a user and a product.


That said, there is a bit of a hack you can do. You could simply ask for a very large number of recommendations for the user of interest, and then check if any of the items that you want to promote are in the result. If they are, then you can bump those to the top of the promotion list.


By the way, examples of calls like this one are documented in the links Nati sent you earlier.


POST https://<service_name>.azurewebsites.net/api/models/e16198c0-3a72-4f4d-b8ab-e4c07c9bccdb/recommend?userId=0001

x-api-key: your_api_key


Note that if you want items in the transaction history of the user to be returned in the recommendations, you need to set the  allowSeedItemsInRecommendations parameter to true when you first create the model.


Thanks!

-Luis

**Comment by dataedgehungary** [comment]
So the problem I am facing now is that the maximum number of recommendations I am allowed to request is 100,

Could I unlock that somehow?

**Comment by LuisCabrer** [comment]
Good point. I guess you could modify the code... and redeploy the solution.

See https://github.com/Microsoft/Product-Recommendations/blob/master/source/Recommendations.WebApp/Controllers/ModelsRecommendController.cs  line 122.

Thanks!
-Luis.

**Comment by filipwastberg** [comment]
Did this solve your problem @dataedgehungary ?
